src/src/pages/capabilities/jsonmetadata/index.js

```diff
@@ -0,0 +1,203 @@
+import SelectedCapabilityContext from "../SelectedCapabilityContext";
+import AppContext from "../../../AppContext";
+import { Button, ButtonStack, Spinner } from "@dfds-ui/react-components";
+import Ajv from "ajv";
+import SyntaxHighlighter from "react-syntax-highlighter";
+import styles from "./jsonmetadata.module.css";
+import PageSection from "../../../components/PageSection";
+import MonacoEditor, { useMonaco } from "@monaco-editor/react";
+import { vs as syntaxStyle } from "react-syntax-highlighter/dist/esm/styles/hljs";
+export function JsonMetadataWithSchemaViewer() {
+  const monaco = useMonaco();
+  const { metadata, setCapabilityJsonMetadata } = useContext(
+    SelectedCapabilityContext,
+  );
+  const [isEditing, setIsEditing] = useState(false);
+  const [isDirty, setIsDirty] = useState(false);
+  const { selfServiceApiClient } = useContext(AppContext);
+  const [currentMetadataString, setCurrentMetadataString] = useState(metadata);
+  const [jsonString, setJsonString] = useState("");
+  const [schemaString, setSchemaString] = useState("");
+  const [validationError, setValidationError] = useState("");
+
+  useEffect(() => {
+    if (monaco && schemaString) {
+      const monacoSchema = {
+        fileMatch: ["*"],
+        schema: JSON.parse(schemaString),
+      };
+      monaco.languages.json.jsonDefaults.setDiagnosticsOptions({
+        validate: true,
+        schemas: [monacoSchema],
+      });
+    }
+  }, [schemaString, monaco]);
+
+  useEffect(() => {
+    if (metadata) {
+      const prettyMetadata = prettifyJsonString(metadata);
+      setCurrentMetadataString(prettyMetadata);
+      setJsonString(prettyMetadata);
+    }
+  }, [metadata]);
+
+  useEffect(() => {
+    if (schemaString === "") {
+      selfServiceApiClient.getCapabilityJsonMetadataSchema().then((schema) => {
+        setSchemaString(prettifyJsonString(schema));
+      });
+    }
+  }, []);
+
+  const checkIfJsonIsParsable = (json) => {
+    try {
+      JSON.parse(json);
+      return true;
+    } catch (exception) {
+      setValidationError("Invalid JSON: " + exception.message);
+      return false;
+    }
+  };
+
+  const checkIfFollowsSchema = (json) => {
+    try {
+      const ajv = new Ajv();
+      const validate = ajv.compile(JSON.parse(schemaString));
+      const valid = validate(JSON.parse(json));
+      if (!valid) {
+        setValidationError("Schema Error: " + validate.errors[0].message);
+      }
+      return valid;
+    } catch (exception) {
+      setValidationError("Schema Error: " + exception.message);
+      return false;
+    }
+  };
+
+  const isJsonValid = (json) => {
+    return validationError === "";
+  };
+
+  const prettifyJsonString = (json) => {
+    return JSON.stringify(JSON.parse(json), null, 2);
+  };
+
+  const submitJsonMetadata = async () => {
+    const prettyJsonString = prettifyJsonString(jsonString);
+    if (isDirty) {
+      setCapabilityJsonMetadata(prettyJsonString);
```

Comment:
Does what we eventually send to the server have to be pretty?

Reply:
Nope, but I don't see any downsides to it, I guess it could be nice for debugging